lib/spack/spack/solver/concretize.lp

```diff
@@ -559,6 +565,22 @@ error(2, "Cannot satisfy requirement group for package '{0}'", Package) :-
 variant(Package, Variant) :- variant_condition(ID, Package, Variant),
                              condition_holds(ID).
 
+% propagate the variant
+variant_value(Descendant, Variant, Value) :-
+  node(Package), path(Package, Descendant),
+  variant(Package, Variant),
+  variant(Descendant, Variant),
+  variant_value(Package, Variant, Value),
+  variant_propagate(Package, Variant),
+  not variant_set(Descendant, Variant),
+  variant_possible_value(Descendant, Variant, Value).
+
+error(2, "{0} and dependency {1} cannot both propagate variant '{2}'", Package, Descendant, Variant) :- 
+  variant_propagate(Package, Variant),
+  variant_propagate(Descendant, Variant),
+  path(Package, Descendant),
+  build(Package).
```

Comment:
@becker33 This is to prevent having the same variant propagate with different values in a spec. Like this:
`spack spec openssl ++docs ^berkeley-db ~~docs`

Reply:
We definitely want to avoid multiple values propagating to the same node. What about

```
error(2, "{0} and {1} propagate incompatible values {2} and {3} for variant {4} to {5}", Package1, Package2, Value1, Value2, Variant, Descendant) :-
  variant_propagate(Package1, Variant),
  variant_propagate(Package2, Variant),
  variant_value(Package1, Variant, Value1),
  variant_value(Package2, Variant, Value2),
  value1 != value2,
  path(Package1, Descendent),
  path(Package2, Descendent),
  build(Descendent).
```

I'm concerned that the current versions prevents `foo ++shared ^bar~~shared ^baz` but it doesn't prevent two unrelated parents propagating to the same place.

For a test, maybe we should have a test that concretizes `dt-diamond ^dt-diamond-right ++shared ^dt-diamond-left ~~shared` and ensure that it throws a reasonable error.

Everything below here should probably be in a separate PR:

In the long run, we'll want to come back to this and enable propagation to be overridden for a sub-DAG. For example, if `foo -> bar -> baz -> corge`, we should be able to do `foo ++shared ^baz ~~shared` and result in

`foo+shared ^bar+shared ^baz~shared ^corge~shared`.

For that, we could have something like "inherit if there's a path and there's not an intermediate package that also propagates this variant in that path". But that's futureware.

Reply:
@becker33 Thanks for catching this. I'll make some changes, but I'm not sure that we need to compare `value1` and `value2`. Shouldn't it fail regardless.?

Reply:
that should be `Value1 != Value2`, I messed up the capitalization. I think we want a check for that, otherwise something like 

`hdf5++shared ^mpich++shared` 

will fail.

It's not the end of the world if that fails, but it seems like it's nearly trivial to catch that case as one we don't mind.

Reply:
@becker33 But wouldn't shared already propagate to `mpich`'s dependencies because of `hdf5++shared`?